smolyay/basis.py

```diff
@@ -403,29 +482,393 @@ def _derivative(self, x):
         float
             Value of the derivative of Trigonometric polynomial.
 
+        n : int, optional
+            order of derivative. Default is 1.
+
         Raises
         ------
         ValueError
             If input is outside the domain `[0, 2\pi]`
         """
         x = numpy.asarray(x)
-        return self.frequency * 1j * numpy.exp(x * self.frequency * 1j)
+        return numpy.exp(x * self.frequency * 1j) * (self.frequency * 1j) ** n
+
+
+class BasisFunctionSet(collections.abc.Sequence):
+    """Set of basis functions and sample points."""
+
+    @abc.abstractmethod
+    def __init__(self):
+        pass
+
+    @property
+    def basis_functions(self):
+        """list: Basis functions."""
+        return self._basis_functions
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain of the `basis_functions`"""
+        return self.basis_functions[0].domain
 
+    def __len__(self):
+        return len(self.basis_functions)
```

Comment:
Prefer to use the internal variables directly if possible, to avoid 1 indirection.
```suggestion
        return len(self._basis_functions)
```